File: step1/leetcode/dp_lc121_buyStock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        #idea
        # f[i] = i-nd 's max profit
        # f[i] = max(f[i-1], prices[i] - min(prices[0:i]))
        # f[0] = 0
        
        if len(prices) == 0:
            return 0
        f = len(prices)*[0]
        
        min_price = prices[0]
        for i in range(1,len(prices)):
            if prices[i] < min_price: 
                min_price = prices[i]
            f[i] = max(f[i-1], prices[i] - min_price)


        return f[len(prices)-1]

File: step1/leetcode/cpp/dp_lc121_buyStock.cpp
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

class Solution {
    public:
    int maxProfit(vector<int>& prices) {
        if (prices.size() == 0){
            return 0;
        }
        vector<int> f(prices.size());
        int min_price = prices[0];
        for(int i = 0; i < prices.size(); i++ ){
            if (i == 0){
                f[i] = 0;
            }
            else {
                if (prices[i]  < min_price){
                    min_price = prices[i];
                }
                f[i] = max(f[i-1], prices[i] - min_price);
                //f[i] = max(f[i-1], prices[i] - *min_element(prices.begin(),prices.begin()+i))
            }
        }
        return f[prices.size()-1];
    }
};
